@default directive accepts invalid values

Add a schema with invalid default value:
```shell
$ defradb client schema add 'type User { name: String, active: Boolean @default(bool: blah)}'
```
It competes without errors. When you create a doc arterwards:
```shell
$ defradb client collection create '{ "name": "John" }'
```
You get a panic:
```shell
panic: interface conversion: interface {} is string, not bool

goroutine 1 [running]:
github.com/sourcenetwork/defradb/client.getBool({0x108e5b540?, 0x14001d5bcd0?})
	github.com/sourcenetwork/defradb/client/document.go:386 +0x94
github.com/sourcenetwork/defradb/client.validateFieldSchema({0x108e5b540, 0x14001d5bcd0}, {{0x14001d71d68, 0x6}, 0x1, {0x10972cfb8, 0x10bd71590}, {0x0, 0x0}, 0x1, ...})
	github.com/sourcenetwork/defradb/client/document.go:291 +0xa58
github.com/sourcenetwork/defradb/client.(*Document).Set(0x140012151e0, {0x14001d71d68, 0x6}, {0x108e5b540, 0x14001d5bcd0})
	github.com/sourcenetwork/defradb/client/document.go:736 +0x53c
github.com/sourcenetwork/defradb/client.(*Document).setDefaultValues(0x140012151e0)
	github.com/sourcenetwork/defradb/client/document.go:778 +0xb8
github.com/sourcenetwork/defradb/client.newEmptyDoc({{{0x1, {0x14001d71ad0, 0x4}}, 0x1, 0x1, {0x14001d82340, 0x3b}, {0x10beea120, 0x0, 0x0}, ...}, ...})
	github.com/sourcenetwork/defradb/client/document.go:109 +0xa8
github.com/sourcenetwork/defradb/client.NewDocFromJSON({_, _, _}, {{{0x1, {0x14001d71ad0, 0x4}}, 0x1, 0x1, {0x14001d82340, 0x3b}, ...}, ...})
	github.com/sourcenetwork/defradb/client/document.go:170 +0x4c
github.com/sourcenetwork/defradb/cli.MakeCollectionCreateCommand.func1(0x14001d65208, {0x14001d5aa80?, 0x1?, 0x107f2b1bc?})
	github.com/sourcenetwork/defradb/cli/collection_create.go:100 +0x26c
github.com/spf13/cobra.(*Command).execute(0x14001d65208, {0x14001d5aa50, 0x1, 0x1})
	github.com/spf13/cobra@v1.8.1/command.go:985 +0x840
github.com/spf13/cobra.(*Command).ExecuteC(0x14001d6db08)
	github.com/spf13/cobra@v1.8.1/command.go:1117 +0x344
github.com/spf13/cobra.(*Command).Execute(0x10bd03388?)
	github.com/spf13/cobra@v1.8.1/command.go:1041 +0x1c
main.main()
	github.com/sourcenetwork/defradb/cmd/defradb/main.go:23 +0x20
```

Note:
With not matching types:
```shell
$ defradb client schema add 'type User { name: String, age: Int @default(int: "1")}'
```
It doesn't panic:
```shell
$ defradb client collection create '{ "name": "John" }'
Error: unexpected type. Property: field, Expected: int64, Actual: string
```
But it needs to be caught on schema creation.

--- a/docs/data_format_changes/i3137-default-value-fix.md
+++ b/docs/data_format_changes/i3137-default-value-fix.md
@@ -1,0 +1,3 @@
+# Default Value Fix
+
+Default value parsing has changed slightly and causes the change detector to fail.

--- a/internal/db/definition_validation.go
+++ b/internal/db/definition_validation.go
@@ -166,6 +166,7 @@ var globalValidators = []definitionValidator{
 	validateSelfReferences,
 	validateCollectionMaterialized,
 	validateMaterializedHasNoPolicy,
+	validateCollectionFieldDefaultValue,
 }
 
 var createValidators = append(
@@ -1013,6 +1014,23 @@ func validateMaterializedHasNoPolicy(
 	for _, col := range newState.collections {
 		if col.IsMaterialized && len(col.QuerySources()) != 0 && col.Policy.HasValue() {
 			return NewErrMaterializedViewAndACPNotSupported(col.Name.Value())
+		}
+	}
+
+	return nil
+}
+
+func validateCollectionFieldDefaultValue(
+	ctx context.Context,
+	db *db,
+	newState *definitionState,
+	oldState *definitionState,
+) error {
+	for name, col := range newState.definitionsByName {
+		// default values are set when a doc is first created
+		_, err := client.NewDocFromMap(map[string]any{}, col)
+		if err != nil {
+			return NewErrDefaultFieldValueInvalid(name, err)
 		}
 	}
 

--- a/internal/db/errors.go
+++ b/internal/db/errors.go
@@ -105,6 +105,7 @@ const (
 	errSelfReferenceWithoutSelf                 string = "must specify 'Self' kind for self referencing relations"
 	errColNotMaterialized                       string = "non-materialized collections are not supported"
 	errMaterializedViewAndACPNotSupported       string = "materialized views do not support ACP"
+	errInvalidDefaultFieldValue                 string = "default field value is invalid"
 )
 
 var (
@@ -679,5 +680,13 @@ func NewErrMaterializedViewAndACPNotSupported(collection string) error {
 	return errors.New(
 		errMaterializedViewAndACPNotSupported,
 		errors.NewKV("Collection", collection),
+	)
+}
+
+func NewErrDefaultFieldValueInvalid(collection string, inner error) error {
+	return errors.New(
+		errInvalidDefaultFieldValue,
+		errors.NewKV("Collection", collection),
+		errors.NewKV("Inner", inner),
 	)
 }

--- a/internal/request/graphql/schema/collection.go
+++ b/internal/request/graphql/schema/collection.go
@@ -427,23 +427,35 @@ func defaultFromAST(
 	if !ok {
 		return nil, NewErrDefaultValueNotAllowed(field.Name.Value, astNamed.Name.Value)
 	}
+	if len(directive.Arguments) != 1 {
+		return nil, NewErrDefaultValueOneArg(field.Name.Value)
+	}
+	arg := directive.Arguments[0]
+	if propName != arg.Name.Value {
+		return nil, NewErrDefaultValueType(field.Name.Value, propName, arg.Name.Value)
+	}
 	var value any
-	for _, arg := range directive.Arguments {
-		if propName != arg.Name.Value {
-			return nil, NewErrDefaultValueInvalid(field.Name.Value, propName, arg.Name.Value)
-		}
-		switch t := arg.Value.(type) {
-		case *ast.IntValue:
-			value = gql.Int.ParseLiteral(arg.Value, nil)
-		case *ast.FloatValue:
-			value = gql.Float.ParseLiteral(arg.Value, nil)
-		case *ast.BooleanValue:
-			value = t.Value
-		case *ast.StringValue:
-			value = t.Value
-		default:
-			value = arg.Value.GetValue()
-		}
+	switch propName {
+	case types.DefaultDirectivePropInt:
+		value = gql.Int.ParseLiteral(arg.Value, nil)
+	case types.DefaultDirectivePropFloat:
+		value = gql.Float.ParseLiteral(arg.Value, nil)
+	case types.DefaultDirectivePropBool:
+		value = gql.Boolean.ParseLiteral(arg.Value, nil)
+	case types.DefaultDirectivePropString:
+		value = gql.String.ParseLiteral(arg.Value, nil)
+	case types.DefaultDirectivePropDateTime:
+		value = gql.DateTime.ParseLiteral(arg.Value, nil)
+	case types.DefaultDirectivePropJSON:
+		value = types.JSONScalarType().ParseLiteral(arg.Value, nil)
+	case types.DefaultDirectivePropBlob:
+		value = types.BlobScalarType().ParseLiteral(arg.Value, nil)
+	}
+	// If the value is nil, then parsing has failed, or a nil value was provided.
+	// Since setting a default value to nil is the same as not providing one,
+	// it is safer to return an error to let the user know something is wrong.
+	if value == nil {
+		return nil, NewErrDefaultValueInvalid(field.Name.Value, propName)
 	}
 	return value, nil
 }

--- a/internal/request/graphql/schema/errors.go
+++ b/internal/request/graphql/schema/errors.go
@@ -30,8 +30,10 @@ const (
 	errPolicyUnknownArgument         string = "policy with unknown argument"
 	errPolicyInvalidIDProp           string = "policy directive with invalid id property"
 	errPolicyInvalidResourceProp     string = "policy directive with invalid resource property"
-	errDefaultValueInvalid           string = "default value type must match field type"
+	errDefaultValueType              string = "default value type must match field type"
 	errDefaultValueNotAllowed        string = "default value is not allowed for this field type"
+	errDefaultValueInvalid           string = "default value is invalid"
+	errDefaultValueOneArg            string = "default value must specify one argument"
 	errFieldTypeNotSpecified         string = "field type not specified"
 )
 
@@ -141,9 +143,24 @@ func NewErrRelationNotFound(relationName string) error {
 	)
 }
 
-func NewErrDefaultValueInvalid(name string, expected string, actual string) error {
+func NewErrDefaultValueOneArg(field string) error {
+	return errors.New(
+		errDefaultValueOneArg,
+		errors.NewKV("Field", field),
+	)
+}
+
+func NewErrDefaultValueInvalid(field string, arg string) error {
 	return errors.New(
 		errDefaultValueInvalid,
+		errors.NewKV("Field", field),
+		errors.NewKV("Arg", arg),
+	)
+}
+
+func NewErrDefaultValueType(name string, expected string, actual string) error {
+	return errors.New(
+		errDefaultValueType,
 		errors.NewKV("Name", name),
 		errors.NewKV("Expected", expected),
 		errors.NewKV("Actual", actual),

--- a/tests/integration/collection_description/with_default_fields_test.go
+++ b/tests/integration/collection_description/with_default_fields_test.go
@@ -58,7 +58,7 @@ func TestCollectionDescription_WithDefaultFieldValues(t *testing.T) {
 							{
 								ID:           3,
 								Name:         "created",
-								DefaultValue: "2000-07-23T03:00:00-00:00",
+								DefaultValue: "2000-07-23T03:00:00Z",
 							},
 							{
 								ID:           4,
@@ -83,6 +83,23 @@ func TestCollectionDescription_WithDefaultFieldValues(t *testing.T) {
 						},
 					},
 				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestCollectionDescription_WithInvalidDefaultFieldValueType_ReturnsError(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: `
+					type Users {
+						active: Boolean @default(bool: invalid)
+					}
+				`,
+				ExpectedError: "default value is invalid. Field: active, Arg: bool",
 			},
 		},
 	}
@@ -116,7 +133,7 @@ func TestCollectionDescription_WithMultipleDefaultFieldValueTypes_ReturnsError(t
 						name: String @default(string: "Bob", int: 10, bool: true, float: 10)
 					}
 				`,
-				ExpectedError: "default value type must match field type. Name: name, Expected: string, Actual: int",
+				ExpectedError: "default value must specify one argument. Field: name",
 			},
 		},
 	}

